xatlas.cpp

```diff
@@ -6426,15 +6426,23 @@ static void triangle_angles(const Vector3 &v1, const Vector3 &v2, const Vector3
 static bool setup_abf_relations(opennl::NLContext *context, int id0, int id1, int id2, const Vector3 &p0, const Vector3 &p1, const Vector3 &p2)
 {
 	// @@ IC: Wouldn't it be more accurate to return cos and compute 1-cos^2?
-	// It does indeed seem to be a little bit more robust.
-	// @@ Need to revisit this more carefully!
+    // @@ MH: sin(acos(v)) is more accurate when v in [-0.25, 0.25] ~ 75-115 deg
+	// @@ outside of this interval sqrt(1 - v^2) is more accurate,
+	// @@ however it can't compensate for the loss of accuracy due to floats :/
+	const float kSineEpsilon = 1e-6;
```

Comment:
What do you think about changing these computations to double precision? The test for zero has to be fixed anyway, of course.

Reply:
I would prefer to solve mesh parametrization in doubles but it would require significant changes in the xatlas library. If we are making big changes to the xatlas we might as well look in a better parametrization algorithm such as: [Fitting to Guidance Gradient Fields](http://mgarland.org/files/papers/param.pdf) which Ignacio Castaño (author of thekla_atlas) suggested in one of his [blog articles ](http://www.ludicon.com/castano/blog/articles/lightmap-parameterization/).

Reply:
Yeah, there are more inspiring posts, like http://www.ludicon.com/castano/blog/2009/12/pate/ :-) But I do agree.

Here I was thinking more about doubling the precision locally for the evaluation of trigonometric functions, divisions, etc. and storing the result in single precision again. But this may not be important after your changes.